tstrait/genetic_value.py

```diff
@@ -0,0 +1,184 @@
+                stack.append(child_node_id)
+            child_node_id = right_sib_array[child_node_id]
+
+    return genotype
+
+
+class GeneticValue:
+    """GeneticValue class to compute genetic values of individuals.
+
+    :param ts: Tree sequence data with mutation.
+    :type ts: tskit.TreeSequence
+    :param trait_df: Pandas dataframe that includes causal site ID, causal allele,
+        simulated effect size, and trait ID.
+    :type effect_size_df: pandas.DataFrame
+    """
+
+    def __init__(self, ts, trait_df):
+        trait_df = trait_df.sort_values(by="site_id")
+        self.ts = ts
+        self.num_trait = len(trait_df.trait_id.unique())
+        self.trait_df = trait_df
+
+    def _individual_genotype(self, tree, site, causal_state, num_nodes):
+        """
+        Returns a numpy array that describes the number of causal mutation in an
+        individual.
+        """
+        has_mutation = np.zeros(num_nodes + 1, dtype=bool)
+        state_transitions = {tree.virtual_root: site.ancestral_state}
+        for m in site.mutations:
+            state_transitions[m.node] = m.derived_state
+            has_mutation[m.node] = True
+        stack = numba.typed.List()
+        for node, state in state_transitions.items():
+            if state == causal_state:
+                stack.append(node)
+
+        if len(stack) == 0:
+            genotype = np.zeros(self.ts.num_individuals)
+        else:
+            genotype = _traversal_genotype(
+                nodes_individual=self.ts.nodes_individual,
+                left_child_array=tree.left_child_array,
+                right_sib_array=tree.right_sib_array,
+                stack=stack,
+                has_mutation=has_mutation,
+                num_individuals=self.ts.num_individuals,
+                num_nodes=num_nodes,
+            )
+
+        return genotype
+
+    def compute_genetic_value_single(self, df):
+        """Computes genetic values of individuals for a single trait.
+
+        This method computes genetic values of individuals based on the simulated
+        effect sizes and the corresponding site ID and causal allele.
+
+        :return: Returns a numpy array of genetic values.
+        :rtype: numpy.ndarray(float)
+        """
+        causal_site_array = df["site_id"]
+        causal_state_array = df["causal_state"]
+        beta_array = df["effect_size"]
+
+        num_nodes = self.ts.num_nodes
+        tree = tskit.Tree(self.ts)
+
+        individual_genetic_array = np.zeros(self.ts.num_individuals)
+
+        for i, single_id in enumerate(causal_site_array):
+            site = self.ts.site(single_id)
+            tree.seek(site.position)
+            individual_genotype = self._individual_genotype(
+                tree=tree,
+                site=site,
+                causal_state=causal_state_array.iloc[i],
+                num_nodes=num_nodes,
+            )
+            individual_genetic_array += individual_genotype * beta_array.iloc[i]
```

Comment:
does it make a difference whether you do this vectorized (for the entire array/dataframe column) in one go vs including this in the for loop?

Reply:
I have a question regarding this. Would you suggest that I create a large genotype matrix through for loop and doing a matrix vector multiplication at the end, instead of doing this in a for loop like this?

Reply:
No I was wrong sorry. Ignore my comment.